sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala

```diff
@@ -108,7 +108,31 @@ abstract class UnsafeProjection extends Projection {
   override def apply(row: InternalRow): UnsafeRow
 }
 
-trait UnsafeProjectionCreator {
+/**
+ * The factory object for `UnsafeProjection`.
+ */
+object UnsafeProjection extends CodegenObjectFactory[Seq[Expression], UnsafeProjection] {
```

Comment:
seems to make it a class is a good idea, we can set default valule like
```
class UnsafeProjection(mode: CodegenObjectFactoryMode = CodegenObjectFactoryMode.currentMode)
```

Reply:
If all subclasses of `CodegenObjectFactory` need `mode`, isn't we ask for it value at `createObject` better?

Reply:
I mean something like
```
abstract class CodegenObjectFactory[IN, OUT] {
  def mode: CodegenObjectFactoryMode

  def createObject(in: IN): OUT = {
    if (mode == ...) ...
    ...
  }
}
```

Reply:
The problem of passing `mode` into class `UnsafeProjection` is, if we create the `UnsafeProjection` object and then change the SQL config, it can't be reflected in ``UnsafeProjection` object's behavior.

If we don't let the config affect this behavior on production, then it should be fine. Then we don't actually need this parameter at most of time, because we only use it in tests. If so, this parameter seems useless. We just need to access `CodegenObjectFactoryMode.currentMode` when `createObject` is called.